### PR TITLE
feat(nexus3): Added license option and integrated HA configuration

### PR DESCRIPTION
**Pull request by timdeluxe**
Latest releases of Nexus allow a "new" way of HA, which i tried to integrate here. It works in our environment (with PSQL as DB and S3 as blob storage).

Note: I thought that the license option/property was needed for HA, but apparently it is saved in the DB, because on startup that property is ignore due to an already available license.
However i left it in the code, because it might be useful to someone.

Let me know what you think.

### CODE REVIEW

File: charts/nexus3/values.yaml
Comment (stevehipwell) [inline charts/nexus3/values.yaml:31]:
```suggestion
replicas: 1
```

File: charts/nexus3/values.yaml
Comment (stevehipwell) [inline charts/nexus3/values.yaml:144]:
```suggestion
  secret:
  key: nexus.license
```
The secret should have a custom key.

File: charts/nexus3/templates/statefulset.yaml
Comment (stevehipwell) [inline charts/nexus3/templates/statefulset.yaml:155]:
```suggestion
              subPath: ${{ .Values.licence.key }}
```

File: charts/nexus3/templates/statefulset.yaml
Comment (stevehipwell) [inline charts/nexus3/templates/statefulset.yaml:11]:
```suggestion
  replicas: {{ .Values.replicas }}
```
This should probably have some more constraints applied.

File: charts/nexus3/templates/deployment.yaml
Comment (stevehipwell) [inline charts/nexus3/templates/deployment.yaml:158]:
```suggestion
```

File: charts/nexus3/templates/deployment.yaml
Comment (stevehipwell) [inline charts/nexus3/templates/deployment.yaml:198]:
```suggestion
```

**Review by stevehipwell — CHANGES_REQUESTED** [review]
@timdeluxe it looks like the only change needed for HA is replicas? If so what are the other configurations which need to be in place for this to work?

File: charts/nexus3/values.yaml
Comment (timdeluxe) [reply charts/nexus3/values.yaml:31]: Do you really think we should just call it replicas? I've left deployment with replicas = 1 hardcoded, because the HA way could just work with statefulset due to the persistent volume - it could work with tmpdir in deployment as well, but f.e. we are using the options to use persistent volumes to keep f.e. logs. Multiple deployment pods won't work with just one ReadWriteOnce PV(C)...

File: charts/nexus3/values.yaml
Comment (timdeluxe) [reply charts/nexus3/values.yaml:144]: will implement this

File: charts/nexus3/templates/deployment.yaml
Comment (timdeluxe) [reply charts/nexus3/templates/deployment.yaml:198]: See https://github.com/stevehipwell/helm-charts/pull/839#discussion_r1419215251

**Comment by timdeluxe** [comment]
> @timdeluxe it looks like the only change needed for HA is replicas? If so what are the other configurations which need to be in place for this to work?

Basically what it listed here: https://help.sonatype.com/repomanager3/product-information/sonatype-nexus-repository-system-requirements/system-requirements-for-high-availability-deployments
In short: PSQL as database, a shared blob storage (S3, Azure Blob Store or NFS), Nexus >= 3.50, a Pro license, a supported docker repo setup (rev-proxy or subdomain connector). All stuff which can't really be checked via the helm chart in my opinion.

**Comment by stevehipwell** [comment]
@timdeluxe how about adding a new `highAvailability` value with the structure set as below? Then it's an explicit opt in, and we'll only support it in `StatefulSet` mode. We could also gate it on the licence being set (I know it's a one time setting but might make it simpler)?

```yaml
highAvailability:
  enabled: false
  replicas: 2
```

File: charts/nexus3/templates/deployment.yaml
Comment (stevehipwell) [reply charts/nexus3/templates/deployment.yaml:158]: The `Deployment` implementation is legacy functionality so I don't add new features there.

File: charts/nexus3/values.yaml
Comment (stevehipwell) [reply charts/nexus3/values.yaml:31]: The `Deployment` would keep it's hard coded replica count, the chart is focused on the `StatefulSet` implementation and `replicas` is the idiomatic value for my charts.

File: charts/nexus3/templates/deployment.yaml
Comment (timdeluxe) [reply charts/nexus3/templates/deployment.yaml:158]: Ok, shouldn't that be mentioned in the README?

**Comment by timdeluxe** [comment]
I tried to implement your proposal. Let me know what you think...

File: charts/nexus3/templates/statefulset.yaml
Comment (stevehipwell) [inline charts/nexus3/templates/statefulset.yaml:14]:
```suggestion
  replicas: 1
```

File: charts/nexus3/templates/statefulset.yaml
Comment (stevehipwell) [inline charts/nexus3/templates/statefulset.yaml:158]:
```suggestion
              name: license
```

File: charts/nexus3/templates/statefulset.yaml
Comment (stevehipwell) [inline charts/nexus3/templates/statefulset.yaml:195]:
```suggestion
        - name: license
```

File: charts/nexus3/README.md
Comment (stevehipwell) [inline charts/nexus3/README.md:81]: The `licence.key` value needs adding here.

**Review by stevehipwell — CHANGES_REQUESTED** [review]
I've added a couple more suggestions, this should be good to merge once they're addressed.

**Comment by timdeluxe** [comment]
@stevehipwell updated the PR now.
Can you check the unresolved threads from your earlier feedback? I had ask you stuff there, which you didn't answer yet. Thanks!

File: charts/nexus3/README.md
Comment (stevehipwell) [inline charts/nexus3/README.md:39]:
```suggestion
```

This isn't a value, but we're missing the HA ones.

**Review by stevehipwell — CHANGES_REQUESTED** [review]
I can't see any unresolved comments. I think we're just a change to the README away from being good to merge.

**Comment by timdeluxe** [comment]
Will update the README, but before i do, i meant this threads (hope the links work):
https://github.com/stevehipwell/helm-charts/pull/839#discussion_r1419213907
https://github.com/stevehipwell/helm-charts/pull/839#discussion_r1419670447

File: charts/nexus3/templates/deployment.yaml
Comment (stevehipwell) [reply charts/nexus3/templates/deployment.yaml]: This is actually the one original chart not to get a SemVer major bump to deprecate the `Deployment` pattern. I'll add something to the README before release to cover recommending use as a `StatefulSet`.

**Comment by stevehipwell** [comment]
@timdeluxe they should be resolved now. You will also need to rebase.

**Comment by timdeluxe** [comment]
Updated README and rebased on current state of main branch

**Comment by stevehipwell** [comment]
@timdeluxe this LGTM. Could you rebase so I can approve and merge?

**Comment by timdeluxe** [comment]
@stevehipwell rebase on what? I already rebased on current state of main?

**Comment by timdeluxe** [comment]
ah well, something got wrong, i am sorry - will do

**Comment by timdeluxe** [comment]
Now it is rebased on your latest commits in main

File: charts/nexus3/templates/configmap-properties.yaml
Comment (stevehipwell) [inline charts/nexus3/templates/configmap-properties.yaml:11]:
```suggestion
    nexus.licenseFile=/nexus-data/etc/nexus.license
```
Sorry I missed this in the review. The `licence.key` is only used in the mapping as we're mounting it to this absolute path.

**Comment by timdeluxe** [comment]
Meh. Looks like i need new glasses as well.
Hopefully now we catched everything?

**Review by stevehipwell — CHANGES_REQUESTED** [review]
@timdeluxe the licence path is still incorrect, but once this is corrected we're good to merge.

**Comment by timdeluxe** [comment]
Done!

**Comment by timdeluxe** [comment]
@stevehipwell Is there a chance that we can get merged and released this soon? It is somewhat work related and i'd like to get it done before xmas vacations, that would be awesome :-)

**Comment by timdeluxe** [comment]
@stevehipwell Hope you had a nice christmas and new years eve... Is it possible to get progress here?

**Review by stevehipwell — APPROVED** [review]
LGTM

**Comment by stevehipwell** [comment]
@timdeluxe sorry this didn't get in before I went on annual leave and had some time away from GitHub. Usually I'd ask for CHANGELOG entries in the PR but in order to get this released I'll do those as part of the release process. I assume that you'd like a chart release before the next Nexus release?

**Comment by timdeluxe** [comment]
> sorry this didn't get in before I went on annual leave and had some time away from GitHub. 

No worries, i hope i did not annoy you too much. Sometimes people tend to "forget" PRs (i include myself there), so i ask from time to time. Thanks for your reply now.

> Usually I'd ask for CHANGELOG entries in the PR but in order to get this released I'll do those as part of the release process. I assume that you'd like a chart release before the next Nexus release?

Chart release before next Nexus release would be awesome, because we currently use my patched version of the chart as a local copy in a pretty quirky terraform infrastructure code setup. Going back to a normal chart release would make things a lot easier for us again.

**Pull request merged**
